docroot/profiles/custom/cgov_site/modules/custom/cgov_core/cgov_core.module

```diff
@@ -979,3 +1003,90 @@ function cgov_core_preprocess_views_view_field(array &$variables) {
   $timestamp = $variables['row']->_entity->getChangedTime();
   $variables['output'] = \Drupal::service('date.formatter')->format($timestamp,'short','',NULL,'en');
 }
+
+/**
+ * Method to determine if we are in batch or CLI
+ */
+function _cgov_core_is_programatic() {
+  $batch = batch_get();
+  $is_batch = !empty($batch['sets']);
+  $is_cli = (PHP_SAPI === 'cli');
+  return ($is_cli || $is_batch);
+}
+
+
+/**
+ * Implements hook_ENTITY_TYPE_update().
+ *
+ * Clears the cache for a path alias when it is updated.
+ * Updates only happen if fields that affect the path are changed.
+ * Updates don't occur until the entity revision is published.
+ */
+function cgov_core_path_alias_update(PathAliasInterface $alias) {
+  // Skip if this is not user triggered.
+  if (_cgov_core_is_programatic()) {
+    return;
+  }
+  $alias_path = $alias->getAlias();
+  $langcode = $alias->language()->getId();
+
+  // Check if language prefix is enabled for the language.
+  // If so, get the language prefix for the URL.
+  $language_manager = \Drupal::service('language_manager');
+  $default_langcode = $language_manager->getDefaultLanguage()->getId();
+  $prefix = '';
+  if ($langcode && $langcode !== $default_langcode) {
+    $prefix = '/' . \Drupal::config('language.negotiation')->get("url.prefixes.$langcode");
+  }
+
+  // Clear the cache for the resulting URL.
+  $url = $prefix . $alias_path;
+  _cgov_core_invalidate_urls([$url]);
+}
+
+
+/**
+ * Implements hook_entity_update().
+ *
+ * Performs a cache clear for the entity's URL on first publish.
+ *
+ * This is required for instances where an entity with a path alias is created
+ * but stays in draft for a while before being published. hook_path_alias_update
+ * won't fire on publish because the path has not changed, but we want to clear the cache.
+ */
+function cgov_core_entity_update(EntityInterface $entity) {
+  // If our entity type is not publishable.
+  if (!$entity instanceof EntityPublishedInterface
+      // If our entity is not published.
+      || !$entity->isPublished()
+      // If our previous moderation state was not "In Review".
+      || (!$entity->moderation_state || $entity->original->moderation_state->value !== "review")
+      // If this was not user triggered.
+      || _cgov_core_is_programatic()) {
```

Comment:
I would move this to its own if at the start as the cheapest exit.

Each if can have a comment.